### PR TITLE
enforce id-based ordering before attempting to obtain partition locks.

### PR DESCRIPTION
**Pull request by emmett9001**
for #253

### CODE REVIEW

File: pykafka/simpleconsumer.py [hunk 1]
```diff
@@ -507,10 +507,13 @@ def _handle_success(parts):
 
         log.info("Resetting offsets for %s partitions", len(list(owned_partition_offsets)))
 
+        # sort offsets to avoid deadlocks
+        sorted_offsets = sorted(iteritems(owned_partition_offsets), key=lambda k: k[0].partition.id)
```
Comment (ltucker) [inline pykafka/simpleconsumer.py]: I think you may want to recompute this within your retry loop since owned_partition_offsets is mutated by the loop.

**Comment by emmett9001** [comment]
@yungchin This could use a sanity check.

**Comment by yungchin** [comment]
@emmett9001 is the sorting in `fetch()` strictly necessary? As that code just skips over any partitions it cannot grab a lock on, doesn't that mean there shouldn't be an issue with deadlocks there? Other than that, this looks good to me.

**Comment by emmett9001** [comment]
No, I don't think it's strictly necessary. I figured sorting is a good idea in general because it makes the whole system more predictable.

**Comment by yungchin** [comment]
Ok, yeah that seems fine to me - I don't imagine the sorting is very expensive.

I've not been able to reproduce the bug here, so can't confirm that it's fixed, but I'm convinced that sorting before taking all the locks is a very real improvement either way.

**Pull request merged**